File: libraries/typescript/src/assetpath.ts
// Lazy materialization of vault assets to a per-handle tempdir.
//
// Some SDKs only accept a filesystem path (GOOGLE_APPLICATION_CREDENTIALS,
// OpenSSL cert paths, …). For those, `assetPath()` writes the asset
// bytes to a 0600 file inside a 0700 per-handle tempdir and returns the
// path. `assetBytes()` (on the Vsync handle) should be the default in
// new code — it never touches the filesystem.
//
// Honest limits: SIGKILL does not run `close()`. A file may leak until
// next reboot (tmpfs) or until a sweep. v0.12 §6.

import { closeSync, existsSync, mkdtempSync, openSync, rmSync, writeSync, chmodSync } from "node:fs";
import { tmpdir } from "node:os";
import { basename, join } from "node:path";

/** On Linux, prefer /dev/shm (tmpfs) so the bytes never hit the platter. */
function preferredTmpdirBase(): string {
  if (process.platform === "linux" && existsSync("/dev/shm")) return "/dev/shm";
  return tmpdir();
}

export class AssetMaterializer {
  private tempdir: string | null = null;
  private cache = new Map<string, string>();
  private closed = false;

  /** Lazily create the per-handle tempdir with mode 0700. */
  private ensureDir(): string {
    if (this.tempdir === null) {
      this.tempdir = mkdtempSync(join(preferredTmpdirBase(), `vsync-${process.pid}-`));
      try {
        // mkdtemp creates with the process umask applied; force 0700.
        chmodSync(this.tempdir, 0o700);
      } catch {
        // Non-POSIX platforms (Windows) may reject; ignore.
      }
    }
    return this.tempdir;
  }

  /**
   * Write `payload` under a sanitised `name` and return its absolute path.
   * Repeat calls with the same name return the cached path; the original
   * bytes are NOT overwritten.
   */
  materialize(name: string, payload: Uint8Array): string {
    if (this.closed) {
      throw new Error("AssetMaterializer: already closed");
    }
    const cached = this.cache.get(name);
    if (cached !== undefined) return cached;

    // Defang the asset name: take basename only so a vault entry like
    // "../../etc/passwd" can't escape the tempdir. v0.12 makes no claim
    // against caller-controlled names (vault is operator-trusted), but
    // containment is the polite default.
    const safe = basename(name) || "_asset";
    const dir = this.ensureDir();
    const p = join(dir, safe);

    // O_WRONLY|O_CREAT|O_TRUNC with explicit 0600 — process umask is
    // irrelevant because we pass mode to open(2) directly.
    const fd = openSync(p, "w", 0o600);
    try {
      // node's writeSync accepts a Uint8Array.
      const buf = Buffer.from(payload.buffer, payload.byteOffset, payload.byteLength);
      writeSync(fd, buf);
    } finally {
      closeSync(fd);
    }
    try {
      // Re-chmod in case the platform's open(2) didn't honour mode exactly
      // (umask gets applied on some setups).
      chmodSync(p, 0o600);
    } catch {
      // Same caveat as ensureDir.
    }
    this.cache.set(name, p);
    return p;
  }

  /** Best-effort cleanup. Idempotent. Failures are swallowed. */
  close(): void {
    if (this.closed) return;
    this.closed = true;
    if (this.tempdir !== null) {
      try {
        rmSync(this.tempdir, { recursive: true, force: true });
      } catch {
        // The process is exiting; the OS will reclaim tmpfs on reboot.
      }
      this.tempdir = null;
    }
    this.cache.clear();
  }
}

File: libraries/go/assetpath.go
package vsync

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

// assetMaterializer owns a per-handle tempdir and lazily writes each asset
// to a 0600 file inside it (v0.12 §6). Repeat calls with the same name
// return the cached path without rewriting.
//
// SIGKILL does not run close(); on tmpfs the dir clears on reboot, on
// disk-backed /tmp the OS / next sweep cleans it up. We document this in
// the README — we don't promise more than the OS gives us.
type assetMaterializer struct {
	tempdir string
	cache   map[string]string
	closed  bool
}

func (a *assetMaterializer) ensureDir() error {
	if a.tempdir != "" {
		return nil
	}
	// On Linux, prefer /dev/shm so the bytes stay off the platter.
	base := ""
	if runtime.GOOS == "linux" {
		if st, err := os.Stat("/dev/shm"); err == nil && st.IsDir() {
			base = "/dev/shm"
		}
	}
	dir, err := os.MkdirTemp(base, fmt.Sprintf("vsync-%d-", os.Getpid()))
	if err != nil {
		return err
	}
	// MkdirTemp uses 0700 on Unix already; chmod for belt-and-braces in
	// case a non-standard umask interfered.
	if err := os.Chmod(dir, 0o700); err != nil {
		os.RemoveAll(dir)
		return err
	}
	a.tempdir = dir
	a.cache = make(map[string]string)
	return nil
}

func (a *assetMaterializer) materialize(name string, payload []byte) (string, error) {
	if a.closed {
		return "", fmt.Errorf("assetMaterializer: already closed")
	}
	if err := a.ensureDir(); err != nil {
		return "", err
	}
	if existing, ok := a.cache[name]; ok {
		return existing, nil
	}
	// Defang the name: take basename so a malicious "../../etc/passwd"
	// can't escape. The bundle is operator-trusted but containment is
	// the polite default.
	safe := filepath.Base(name)
	if safe == "" || safe == "." || safe == ".." || safe == string(filepath.Separator) {
		safe = "_asset"
	}
	path := filepath.Join(a.tempdir, safe)
	// O_CREATE|O_WRONLY|O_TRUNC with explicit 0600 — umask of the
	// process is irrelevant when we pass the mode to open(2) directly.
	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
	if err != nil {
		return "", err
	}
	if _, err := f.Write(payload); err != nil {
		f.Close()
		return "", err
	}
	if err := f.Close(); err != nil {
		return "", err
	}
	// Belt-and-braces re-chmod (some platforms apply umask to open's mode).
	if err := os.Chmod(path, 0o600); err != nil {
		return "", err
	}
	a.cache[name] = path
	return path, nil
}

func (a *assetMaterializer) close() {
	if a.closed {
		return
	}
	a.closed = true
	if a.tempdir != "" {
		os.RemoveAll(a.tempdir)
		a.tempdir = ""
	}
	a.cache = nil
}
